### PR TITLE
Allow command to use customized processors

### PR DESCRIPTION
**Pull request by jderusse**
This PR allow users to defines they consumer middleware stack

Try to keep BC by converting `processor_stack` into `consumer.middleware_stack` and allow custom className as parameters.

```
swarrot:
    consumers:
        my_consumer:
            processor: app.my_processor
            queue: queue_name
            middleware_stack: # order matter
                # - configurator: swarrot.processor.insomniac
                - configurator: swarrot.processor.signal_handler
                  # extras:
                  #     signal_handler_signals:
                  #         - SIGTERM
                  #         - SIGINT
                  #         - SIGQUIT
                - configurator: swarrot.processor.max_messages
                  # extras:
                  #     max_messages: 100
                - configurator: swarrot.processor.max_execution_time
                  # extras:
                  #     max_execution_time: 300
                - configurator: swarrot.processor.memory_limit
                  # extras:
                  #     memory_limit: null
                - configurator: swarrot.processor.doctrine_connection
                  # extras:
                  #     doctrine_ping: true
                  #     doctrine_close_master: true
                - configurator: swarrot.processor.doctrine_object_manager
                - configurator: swarrot.processor.exception_catcher

                - configurator: swarrot.processor.ack
                  # extras:
                  #     requeue_on_error: false
                - configurator: swarrot.processor.retry
                  # extras:
                  #     retry_exchange: retry
                  #     retry_attempts: 3
                  #     retry_routing_key_pattern: 'retry_%%attempt%%'
                - configurator: swarrot.processor.instant_retry
                  # extras:
                  #     instant_retry_delay: 2000000
                  #     instant_retry_attempts: 3
                # - configurator: swarrot.processor.new_relic
                #   extras:
                #       new_relic_app_name: ~
                #       new_relic_license: ~
                #       new_relic_transaction_name: ~

                # - configurator: swarrot.processor.rpc_client
                #   extras:
                #       rpc_client_correlation_id: ~
                # - configurator: swarrot.processor.rpc_server
```

should fixe #55 and #81 

TODO:
* [x] Add tests

### CODE REVIEW

File: Command/SwarrotCommand.php [hunk 1]
```diff
@@ -58,99 +55,47 @@ protected function configure()
...
+            $optionsResolver->setOptional(['queue', 'connection']);
         }
 
+        $factory = $this->getContainer()->get('swarrot.factory.default');
+        $messageProvider = $factory->getMessageProvider($options['queue'], $options['connection']);
+
         $consumer = new Consumer($messageProvider, $processor, $optionsResolver);
 
-        $consumer->consume($this->getOptions($input));
+        dump($options);die;
```
Comment (odolbeau) [inline Command/SwarrotCommand.php]: <3

File: Command/SwarrotCommand.php [hunk 2]
```diff
@@ -162,36 +107,19 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
-                $key = $this->extras['retry_routing_key_pattern'];
-            }
-            $key = str_replace('%queue%', $input->getArgument('queue'), $key);
-            $options['retry_key_pattern'] = $key;
-
-            $attempts = 3;
-            if ($input->hasOption('retry-attempts')) {
-                $attempts = (int) $input->getOption('retry-attempts');
+        $options = $this->extras + [
+                'queue' => $input->getArgument('queue'),
```
Comment (odolbeau) [inline Command/SwarrotCommand.php]: /!\ alignment

File: DependencyInjection/SwarrotExtension.php [hunk 3]
```diff
@@ -99,17 +99,66 @@ public function getNamespace()
      */
     public function buildCommand(ContainerBuilder $container, $name, array $consumerConfig, array $processorStack)
     {
+        // Keep BC by converting $processorStack into processorConfigurators
```
Comment (odolbeau) [inline DependencyInjection/SwarrotExtension.php]: Shouldn't it be done inside the Configuration class?

**Comment by odolbeau** [comment]
Looks good! :)
I'll try to add more tests on the master branch in order to ensure that we can release this awesome feature without BC break. :)

File: Command/SwarrotCommand.php (hunk 1 above)
Comment (jderusse) [inline Command/SwarrotCommand.php]: this explains a lot of things ^^

File: DependencyInjection/SwarrotExtension.php [hunk 4]
```diff
@@ -99,17 +99,66 @@ public function getNamespace()
...
+            'retry' => 'swarrot.processor.retry',
+            'signal_handler' => 'swarrot.processor.signal_handler',
+        ];
+        foreach ($map as $stackKey => $configurator) {
+            if (array_key_exists($stackKey, $processorStack)) {
+                // Replace the className (only for BC purpose)
+                $container->findDefinition($configurator)->replaceArgument(0, $processorStack[$stackKey]);
+                // Inject the whole extra in each processorConfigurators because in legacy alle extras was mixed
+                $consumerConfig['middleware_stack'][$configurator] = [
+//                    'extras' => $consumerConfig['extras'],
```
Comment (jderusse) [inline DependencyInjection/SwarrotExtension.php]: should be uncommented

File: DependencyInjection/SwarrotExtension.php (hunk 3 above)
Comment (jderusse) [inline DependencyInjection/SwarrotExtension.php]: Yep, you're right !

**Comment by jderusse** [comment]
Comments take into account and add some tests.
But, indeed, we need more functional tests.

**Comment by odolbeau** [comment]
Sorry for the lack of reactivity. :/
I'll merge #76 tomorrow and take time to add new tests before merging this PR. Sorry for the time it takes.

**Comment by jderusse** [comment]
What about using travis' docker feature to run functional tests with a running rabbitmq server?

Si the acmePHP symfony-bundle:
https://github.com/acmephp/symfony-bundle/blob/master/.travis.yml#L39
https://github.com/acmephp/symfony-bundle/blob/master/features/generate.feature#L3-L16

**Comment by odolbeau** [comment]
Can you confirmed that your PR also fix #64 (ping @metfan)?

Anyway, PR #76 merged.
Can you rebase your branch please?

**Comment by metfan** [comment]
Yes @odolbeau it's fixed #64.
Awesome job @jderusse. 

I'm asking myself why you create traits ProcessorConfiguratorEnableAware and ProcessorConfiguratorExtrasAware instead of abstract base class? They are not optionable behaviors cause they are defined in interface ProcessorConfiguratorInterface.

**Comment by jderusse** [comment]
@metfan Good point, I'd want to split the interface and forget to do it. BTW I don't use inheritance because it's not a good way to factorize code. 

I push a new commit, to segregate the "enable" feature and rebase the branch on top of Master.

Given there is no functional tests, do you test it in real conditions ?

File: Tests/DependencyInjection/ConfigurationTest.php [hunk 5]
```diff
@@ -16,6 +16,9 @@ public function test_it_is_initializable()
         );
     }
 
+    /**
+     * @group legacy
+     */
     public function test_with_default_configuration()
```
Comment (odolbeau) [inline Tests/DependencyInjection/ConfigurationTest.php]: Is it possible to rename this test with something like `test_with_old_default_configuration` (or something better, I'm not very inspired) and create a new one to test the new default configuration (without any deprecated key) generated by symfony?

**Comment by odolbeau** [comment]
I hope I'll have time to test this PR this week. Sorry for the delay. :/

**Comment by odolbeau** [comment]
Finally tested!
Seems everything works fine, thanks again for your contrbution. :)

**Comment by odolbeau** [comment]
I tested it too quickly... At least I didn't create a tag yet!
I have an issue with the `isEnable` method which is called for each configurator in the command even if they don't implement it.
I can't find any use case where the `isEnable` method is not required. I'm thinking about removing the  `ProcessorConfiguratorEnablableInterface` to move the `isEnabled` method directly inside the `ProcessorConfiguratorInterface`.
What do you think?

**Comment by jderusse** [comment]
IsEnabled works for some processor who expose an option to deactivate it. (Like --no-ack in AckProcessor). But some processor don't expose such feature (like Insomniac or InstantRetry). Indeed, I should check the interface before calling the method

**Comment by odolbeau** [comment]
Isn't simpler to keep only one interface & to use the appropriate trait in all processors which can't be disabled for now?

**Pull request merged**